cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_test.go

```diff
@@ -0,0 +1,156 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/fake"
+)
+
+func TestGetAvailabelPackagesBadConfig(t *testing.T) {
+	testCases := []struct {
+		name         string
+		clientGetter func(context.Context) (dynamic.Interface, error)
+		statusCode   codes.Code
+	}{
+		{
+			name:         "returns internal error when no getter configured",
+			clientGetter: nil,
+			statusCode:   codes.Internal,
+		},
+		{
+			name: "returns unknown with the original error message when configGetter errors",
+			clientGetter: func(context.Context) (dynamic.Interface, error) {
+				return nil, fmt.Errorf("Bang!")
+			},
+			statusCode: codes.FailedPrecondition,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			s := Server{clientGetter: tc.clientGetter}
+
+			_, err := s.GetAvailablePackages(context.Background(), &corev1.GetAvailablePackagesRequest{})
+
+			if err == nil {
+				t.Fatalf("got: nil, want: error")
+			}
+
+			if got, want := status.Code(err), tc.statusCode; got != want {
+				t.Errorf("got: %+v, want: %+v", got, want)
+			}
+		})
+	}
+
+}
+
+func packageFromSpec(spec map[string]interface{}) *unstructured.Unstructured {
+	return &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "package.carvel.dev/v1alpha1",
+			"kind":       "Package",
+			"metadata": map[string]interface{}{
+				"name": fmt.Sprintf("%s.%s", spec["publicName"], spec["version"]),
+			},
+			"spec": spec,
+		},
+	}
+}
+
+func packagesFromSpecs(specs []map[string]interface{}) []*unstructured.Unstructured {
+	pkgs := []*unstructured.Unstructured{}
+	for _, s := range specs {
+		pkgs = append(pkgs, packageFromSpec(s))
+	}
+	return pkgs
+}
+
+func TestGetAvailablePackages(t *testing.T) {
+	testCases := []struct {
+		name             string
+		packageSpecs     []map[string]interface{}
+		expectedPackages []*corev1.AvailablePackage
+	}{
+		{
+			name: "it returns carvel packages from the cluster",
+			packageSpecs: []map[string]interface{}{
+				{
+					"publicName": "tetris.foo.example.com",
+					"version":    "1.2.3",
+				},
+				{
+					"publicName": "another.foo.example.com",
+					"version":    "1.2.5",
+				},
+			},
+			expectedPackages: []*corev1.AvailablePackage{
+				{
+					Name:    "another.foo.example.com",
+					Version: "1.2.5",
+				},
+				{
+					Name:    "tetris.foo.example.com",
+					Version: "1.2.3",
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			pkgs := packagesFromSpecs(tc.packageSpecs)
+			s := Server{
+				clientGetter: func(context.Context) (dynamic.Interface, error) {
+					return fake.NewSimpleDynamicClientWithCustomListKinds(
+						runtime.NewScheme(),
+						map[schema.GroupVersionResource]string{
+							{Group: "package.carvel.dev", Version: "v1alpha1", Resource: "packages"}: "PackageList",
+						},
+						// TODO: Not clear why
+						// pkgs...
+						// won't work.
```

Comment:
Mmm, that's true. It has something to do with the `DeepCopyObject` or sth similar.

I've found an example of that kind of testing in the k8s core (autoscaler), but they are using `runtime.Object` instead of `Unstructured`. 
https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go

Reply:
Thanks Antonio. Note that the issue here was that it worked fine with ` pkgs[0], pkgs[1]` manually added as the arguments, but not when using the `...` to [expand the slice multiple function arguments](https://golang.org/ref/spec#Passing_arguments_to_..._parameters), with `pkgs...`.

`runtime.Object` is the interface, which is implemented by `unstructured.Unstructured` which is why it works with the individual args, but I'm just not sure why it doesn't work when spreading the slice. Anyway, I could work around it by updating the `packagesFromSpecs` to return a slice of the interface type instead. Yay.